### PR TITLE
Improve the procedure for setting the solution

### PR DESCRIPTION
**Pull request by bwspenc**
ref #8763

### CODE REVIEW

**Comment by permcody** [comment]
Wait, does this fix that test? I just put up a PR to skip it.

**Comment by bwspenc** [comment]
@permcody I think the problems with getting the unique_id may have originated from the code that this replaces, so I'm hopeful that this might fix that flaky test.  Regardless, this should be far more efficient because it just involves a single loop over the nodes and the elements.

**Comment by permcody** [comment]
So what are we doing with this PR? Are you going to work on it more, or would you like it reviewed and merged?

**Comment by bwspenc** [comment]
@permcody I'd like this to be reviewed and merged.  This is a better way to do things regardless of that test.

File: modules/xfem/include/base/XFEM.h [hunk 1]
```diff
@@ -270,6 +267,20 @@ class XFEM : public XFEMInterface
...
   /**
+   * Set the solution for a set of DOFs
+   * @param stored_solution      Stored solution values to set the solution to
+   * @param stored_solution_dofs Dof indices for the entries in stored_solution
+   * @param current_solution     Current solution vector that will be set
+   * @param old_solution         Old solution vector that will be set
+   * @param older_solution       Older solution vector that will be set
+   */
+  void setSolutionForDOFs(const std::vector<Real> & stored_solution,
+                          const std::vector<unsigned int> & stored_solution_dofs,
```
Comment (permcody) [inline modules/xfem/include/base/XFEM.h]: `unsigned int` -> `dof_id_type`

File: modules/xfem/src/base/XFEM.C [hunk 2]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
                   NumericVector<Number> & old_solution,
                   NumericVector<Number> & older_solution)
 {
-  for (auto dit = stored_solution.begin(); dit != stored_solution.end(); ++dit)
+  for (auto node_it = _mesh->local_nodes_begin(); node_it != _mesh->nodes_end(); ++node_it)
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: I'd move the `nodes_end()` call out of the loop. The libMesh iterators can be slow to evaluate.

File: modules/xfem/src/base/XFEM.C [hunk 3]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+    Node * node = *node_it;
+    auto mit = stored_solution.find(node->unique_id());
+    if (mit != stored_solution.end())
     {
-      elem = getElemFromUniqueID(dit->first);
-      if (!elem)
-        mooseError("Cannot get node or element from unique id");
-      dof_obj_pid = elem->processor_id();
+      const std::vector<Real> & stored_node_solution = mit->second;
+      std::vector<unsigned int> stored_solution_dofs = getNodeSolutionDofs(node, sys);
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: `dof_id_type`

File: modules/xfem/src/base/XFEM.C [hunk 4]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+      setSolutionForDOFs(stored_node_solution,
+                         stored_solution_dofs,
+                         current_solution,
+                         old_solution,
+                         older_solution);
     }
+  }
 
-    if (dof_obj_pid == _mesh->processor_id())
+  for (auto elem_it = _mesh->local_elements_begin(); elem_it != _mesh->local_elements_end();
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Move end iterator out of loop

File: modules/xfem/src/base/XFEM.C [hunk 5]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+  {
+    Elem * elem = *elem_it;
+    auto mit = stored_solution.find(elem->unique_id());
+    if (mit != stored_solution.end())
     {
-      const std::vector<Real> & dof_obj_solution = dit->second;
-      std::vector<unsigned int> stored_solution_dofs =
-          (node ? getNodeSolutionDofs(node, sys) : getElementSolutionDofs(elem, sys));
+      const std::vector<Real> & stored_elem_solution = mit->second;
+      std::vector<unsigned int> stored_solution_dofs = getElementSolutionDofs(elem, sys);
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: `dof_id_type`

File: modules/xfem/src/base/XFEM.C [hunk 6]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+  }
+}
 
-      // Solution vector is stored first for current, then old and older solutions.
-      // These are the offsets to the beginning of the old and older solutions in the vector.
-      const unsigned int old_solution_offset = stored_solution_dofs.size();
-      const unsigned int older_solution_offset = old_solution_offset * 2;
+void
+XFEM::setSolutionForDOFs(const std::vector<Real> & stored_solution,
+                         const std::vector<unsigned int> & stored_solution_dofs,
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: `dof_id_type`

File: modules/xfem/src/base/XFEM.C [hunk 7]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+                         NumericVector<Number> & current_solution,
+                         NumericVector<Number> & old_solution,
+                         NumericVector<Number> & older_solution)
+{
+  // Solution vector is stored first for current, then old and older solutions.
+  // These are the offsets to the beginning of the old and older solutions in the vector.
+  const unsigned int old_solution_offset = stored_solution_dofs.size();
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: wrong type here too - It'll be a `std::size_t`, you might just use auto.

File: modules/xfem/src/base/XFEM.C [hunk 8]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
+  const unsigned int older_solution_offset = old_solution_offset * 2;
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Same here

File: modules/xfem/src/base/XFEM.C [hunk 9]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
 
-      for (unsigned int i = 0; i < stored_solution_dofs.size(); ++i)
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Pedantic, but wrong type

File: modules/xfem/src/base/XFEM.C [hunk 10]
```diff
@@ -1733,41 +1695,59 @@ XFEM::setSolution(SystemBase & sys,
...
-      {
-        current_solution.set(stored_solution_dofs[i], dof_obj_solution[i]);
-        if (_fe_problem->isTransient())
-        {
-          old_solution.set(stored_solution_dofs[i], dof_obj_solution[old_solution_offset + i]);
-          older_solution.set(stored_solution_dofs[i], dof_obj_solution[older_solution_offset + i]);
-        }
-      }
+  for (unsigned int i = 0; i < stored_solution_dofs.size(); ++i)
```
Comment (permcody) [inline modules/xfem/src/base/XFEM.C]: Wrong type

**Review by permcody — CHANGES_REQUESTED** [review]
A few types are wrong. The dof_id_type in particular we need to get right or it breaks the 64-bit indices build.

**Comment by bwspenc** [comment]
@permcody Thanks for the review. I changed those types as you suggested.

Are we running the tests in a configuration where dof_id_type would have been different from unsigned int (maybe with 64 bit indices)?  I'm just wondering if that might be why that test is flaky.

**Comment by permcody** [comment]
We have a 64-bit build. If you look through the list of targets, you'll see it. It does break with some regularity due to improper use of types but it's not part of the tests we run on a PR so no, not this particular issue.

**Review by permcody — APPROVED** [review]

**Pull request merged**